File: app/modules/production/schemas/trace.py
"""溯源 API 契约。合并场景下谱系是 DAG，用 批次列表 + 谱系边列表 表达。"""

import uuid
from datetime import datetime

from pydantic import BaseModel


class TraceExecutionBrief(BaseModel):
    node_name: str
    status: str
    owner_name: str | None
    started_at: datetime
    finished_at: datetime | None
    is_deviation: bool
    abnormal_count: int


class TraceBatch(BaseModel):
    id: uuid.UUID
    batch_no: str
    product_id: uuid.UUID
    status: str
    quantity: float | None
    unit: str | None
    executions: list[TraceExecutionBrief] = []


class TraceLink(BaseModel):
    parent_batch_id: uuid.UUID
    child_batch_id: uuid.UUID
    edge_id: uuid.UUID | None
    allocated_qty: float | None
    is_deviation: bool


class TraceOut(BaseModel):
    root_batch_id: uuid.UUID
    batches: list[TraceBatch]
    links: list[TraceLink]

File: frontend/src/types/production/trace.ts
export interface TraceExecutionBrief {
  node_name: string
  status: string
  owner_name: string | null
  started_at: string
  finished_at: string | null
  is_deviation: boolean
  abnormal_count: number
}

export interface TraceBatch {
  id: string
  batch_no: string
  product_id: string
  product_name: string | null
  status: string
  quantity: number | null
  unit: string | null
  current_stage_name: string | null
  executions: TraceExecutionBrief[]
}

export interface TraceLink {
  parent_batch_id: string
  child_batch_id: string
  /** lineage = 谱系边（derive/merge）；material = 物料边（投料消耗→产出批次） */
  link_type: string
  // lineage 专属
  edge_id: string | null
  allocated_qty: number | null
  is_deviation: boolean
  // material 专属
  intermediate_type_id: string | null
  intermediate_type_name: string | null
  intermediate_batch_no: string | null
  quantity: number | null
  unit: string | null
}

export interface TraceData {
  root_batch_id: string
  batches: TraceBatch[]
  links: TraceLink[]
}
